File: ascii.lua
-- Creator: Alicia Suya Firmansyah
-- Copyright (c) 2025 Alicia Suya Firmansyah. All rights reserved.

local ascii = {}

--- Encode a string to ASCII values
-- @param code The input string to encode
-- @param separator Optional separator for encoded values (defaults to space)
-- @return String of ASCII values or throws error for invalid input
ascii.encode = function(code, separator)
	-- Validate input type
	if type(code) ~= "string" then
		error("Input must be a string, got " .. type(code), 2)
	end
	-- Handle empty string
	if code == "" then
		return ""
	end
	-- Default separator to space if not provided or empty
	separator = (separator and separator ~= "") and separator or " "
	-- Convert each character to its ASCII value
	local tbl = {}
	for i = 1, #code do
		local byte = string.byte(code, i)
		if byte > 255 then
			error("Character at position " .. i .. " is outside valid ASCII range (0-255)", 2)
		end
		tbl[i] = byte
	end
	return table.concat(tbl, separator)
end

--- Decode ASCII values to a string
-- @param code String of space-separated ASCII values or table of numbers
-- @return Decoded string or throws error for invalid input
ascii.decode = function(code)
	if not code then
		return ""
	end
	local res = ""
	if type(code) == "string" then
		-- Handle empty string
		if code == "" then
			return ""
		end
		-- Split by any non-digit separator and convert to characters
		for v in code:gmatch("%d+") do
			local num = tonumber(v)
			if not num or num < 0 or num > 255 then
				error("Invalid ASCII value: " .. v, 2)
			end
			res = res .. string.char(num)
		end
	elseif type(code) == "table" then
		-- Convert table of numbers to characters
		for i, v in ipairs(code) do
			if type(v) ~= "number" or v < 0 or v > 255 then
				error("Invalid ASCII value at index " .. i .. ": " .. tostring(v), 2)
			end
			res = res .. string.char(v)
		end
	else
		error("Input must be a string or table, got " .. type(code), 2)
	end
	return res
end

-- Command-line interface
if arg and #arg >= 2 then
	local command = arg[1]:lower()
	local input = arg[2]
	local separator = arg[3] or " "
	local status, res = pcall(function()
		-- Check if input is a file (ends with .lua)
		if input:match("%.txt$") then
			local file = io.open(input, "r")
			if not file then
				error("Failed to open file: " .. input)
			end
			local content = file:read("*a")
			file:close()

			if command == "enc" then
				return ascii.encode(content, separator)
			elseif command == "dec" then
				-- Handle potential table input from file
				if content:match("^%s*{.+}%s*$") then
					local tbl = load("return " .. content)()
					if type(tbl) ~= "table" then
						error("Invalid table format in file")
					end
					return ascii.decode(tbl)
				else
					return ascii.decode(content)
				end
			else
				error("Unknown command: " .. command .. ". Use 'enc' or 'dec'")
			end
		else
			-- Original behavior for non-file input
			if command == "enc" then
				return ascii.encode(input, separator)
			elseif command == "dec" then
				-- Handle potential table input from command line
				if input:match("^%s*{.+}%s*$") then
					local tbl = load("return " .. input)()
					if type(tbl) ~= "table" then
						error("Invalid table format")
					end
					return ascii.decode(tbl)
				else
					return ascii.decode(input)
				end
			else
				error("Unknown command: " .. command .. ". Use 'enc' or 'dec'")
			end
		end
	end)

	if status then
		print(res)
	else
		print("Error: " .. res)
		os.exit(1)
	end
else
	return ascii
end

File: ascii.js
// Creator: Fellicia Faicha Theda
// Copyright (c) 2025 Fellicia Faicha Theda. All rights reserved.

const ascii = {};

/**
 * Encode a string to ASCII values
 * @param {string} code - The input string to encode
 * @param {string} [separator=" "] - Optional separator for encoded values (defaults to space)
 * @returns {string} String of ASCII values or throws error for invalid input
 */
ascii.encode = function(code, separator = " ") {
    // Validate input type
    if (typeof code !== "string") {
        throw new Error(`Input must be a string, got ${typeof code}`);
    }
    // Handle empty string
    if (code === "") {
        return "";
    }
    // Default separator to space if not provided or empty
    separator = separator && separator !== "" ? separator : " ";
    // Convert each character to its ASCII value
    const result = [];
    for (let i = 0; i < code.length; i++) {
        const byte = code.charCodeAt(i);
        if (byte > 255) {
            throw new Error(`Character at position ${i + 1} is outside valid ASCII range (0-255)`);
        }
        result.push(byte);
    }
    return result.join(separator);
};

/**
 * Decode ASCII values to a string
 * @param {string|number[]} code - String of space-separated ASCII values or array of numbers
 * @returns {string} Decoded string or throws error for invalid input
 */
ascii.decode = function(code) {
    if (!code) {
        return "";
    }
    let result = "";
    if (typeof code === "string") {
        // Handle empty string
        if (code === "") {
            return "";
        }
        // Split by any non-digit separator and convert to characters
        const values = code.match(/\d+/g) || [];
        for (const v of values) {
            const num = parseInt(v);
            if (isNaN(num) || num < 0 || num > 255) {
                throw new Error(`Invalid ASCII value: ${v}`);
            }
            result += String.fromCharCode(num);
        }
    } else if (Array.isArray(code)) {
        // Convert array of numbers to characters
        for (let i = 0; i < code.length; i++) {
            const v = code[i];
            if (typeof v !== "number" || v < 0 || v > 255) {
                throw new Error(`Invalid ASCII value at index ${i + 1}: ${v}`);
            }
            result += String.fromCharCode(v);
        }
    } else {
        throw new Error(`Input must be a string or array, got ${typeof code}`);
    }
    return result;
};

// Command-line interface (Node.js)
if (typeof process !== "undefined" && process.argv.length >= 4) {
    const fs = require("fs").promises;
    const command = process.argv[2].toLowerCase();
    const input = process.argv[3];
    const separator = process.argv[4] || " ";

    (async () => {
        try {
            let result;
            // Check if input is a file (ends with .txt)
            if (input.endsWith(".txt")) {
                try {
                    const content = await fs.readFile(input, "utf8");
                    if (command === "enc") {
                        result = ascii.encode(content, separator);
                    } else if (command === "dec") {
                        // Handle potential array input from file
                        if (content.match(/^\s*\[.*\]\s*$/)) {
                            const array = JSON.parse(content);
                            if (!Array.isArray(array)) {
                                throw new Error("Invalid array format in file");
                            }
                            result = ascii.decode(array);
                        } else {
                            result = ascii.decode(content);
                        }
                    } else {
                        throw new Error(`Unknown command: ${command}. Use 'enc' or 'dec'`);
                    }
                } catch (fileError) {
                    throw new Error(`Failed to open file: ${input}`);
                }
            } else {
                // Original behavior for non-file input
                if (command === "enc") {
                    result = ascii.encode(input, separator);
                } else if (command === "dec") {
                    // Handle potential array input from command line
                    if (input.match(/^\s*\[.*\]\s*$/)) {
                        const array = JSON.parse(input);
                        if (!Array.isArray(array)) {
                            throw new Error("Invalid array format");
                        }
                        result = ascii.decode(array);
                    } else {
                        result = ascii.decode(input);
                    }
                } else {
                    throw new Error(`Unknown command: ${command}. Use 'enc' or 'dec'`);
                }
            }
            console.log(result);
        } catch (error) {
            console.error(`Error: ${error.message}`);
            process.exit(1);
        }
    })();
} else {
    window.ascii = ascii;
}